crates/sui-core/src/consensus_throughput_calculator.rs

```diff
@@ -0,0 +1,517 @@
+
+        // If it's the very first observation we just use it as timestamp and don't count any transactions.
+        let num_of_transactions = if !inner.observations.is_empty() {
+            num_of_transactions
+        } else {
+            0
+        };
+
+        if let Some((front_ts, transactions)) = inner.observations.pop_front() {
+            // First check that the timestamp is monotonically incremented - ignore any observation that is not
+            // later from previous one (it shouldn't really happen).
+            if timestamp_secs < front_ts {
+                warn!("Ignoring observation of transactions:{} as has earlier timestamp than last observation {}s < {}s", num_of_transactions, timestamp_secs, front_ts);
+                return;
+            }
+
+            // Not very likely, but if transactions refer to same second we add to the last element.
+            if timestamp_secs == front_ts {
+                inner
+                    .observations
+                    .push_front((front_ts, transactions + num_of_transactions));
+            } else {
+                inner.observations.push_front((front_ts, transactions));
+                inner
+                    .observations
+                    .push_front((timestamp_secs, num_of_transactions));
+            }
+        } else {
+            inner
+                .observations
+                .push_front((timestamp_secs, num_of_transactions));
+        }
+
+        // update total number of transactions in the observations list
+        inner.total_transactions = inner.total_transactions.saturating_add(num_of_transactions);
+
+        // If we have more values on our window of max values, remove the last one, and update the num of transactions
+        // We also update the traffic profile when we have at least observations_window values in our observations.
+        if inner.observations.len() as u64 > self.observations_window {
+            let (last_element_ts, last_element_transactions) =
+                inner.observations.pop_back().unwrap();
+            inner.total_transactions = inner
+                .total_transactions
+                .saturating_sub(last_element_transactions);
+
+            // get the first element's timestamp to calculate the transaction rate
+            let (first_element_ts, _first_element_transactions) = inner
+                .observations
+                .front()
+                .expect("There should be at least on element in the list");
+
+            let period = first_element_ts.saturating_sub(last_element_ts);
+
+            if period > 0 {
+                let current_throughput = inner.total_transactions / period;
+
+                self.update_traffic_profile(current_throughput, timestamp_secs);
+            } else {
+                warn!("Skip calculating throughput as time period is {}. This is very unlikely to happen, should investigate.", period);
```

Comment:
This can only happen when `observations_window` is set to 0, right?

Reply:
Practically yes, given the bucketing guarantees - well in practice we can't even set `observations_window = 0` as it's a non zero u64 .